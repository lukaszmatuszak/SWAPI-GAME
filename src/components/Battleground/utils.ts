import { Status } from '../../interfaces/Status';
import { Starship } from '../../interfaces/Starship';
import { Character } from '../../interfaces/Character';

export const defineStatusForStarship = (
    firstChallenger: Starship,
    secondChallenger: Starship,
): Status[] => {
    if (Number(firstChallenger.crew) === Number(secondChallenger.crew)) {
        return [Status.DRAW, Status.DRAW];
    }

    return Number(firstChallenger.crew) > Number(secondChallenger.crew)
        ? [Status.WINNER, Status.LOOSER]
        : [Status.LOOSER, Status.WINNER];
};

export const defineStatusForCharacter = (
    firstChallenger: Character,
    secondChallenger: Character,
): Status[] => {
    if (Number(firstChallenger.mass) === Number(secondChallenger.mass)) {
        return [Status.DRAW, Status.DRAW];
    }

    return Number(firstChallenger.mass) > Number(secondChallenger.mass)
        ? [Status.WINNER, Status.LOOSER]
        : [Status.LOOSER, Status.WINNER];
};
