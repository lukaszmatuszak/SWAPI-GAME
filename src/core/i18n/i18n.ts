import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import LanguageDetector from 'i18next-browser-languagedetector';

import locales from './locales/index';

i18n.use(initReactI18next)
    .use(LanguageDetector)
    .init({
        ns: ['Common'],
        defaultNS: 'Common',
        resources: locales,
        fallbackLng: 'en',
        interpolation: {
            escapeValue: false,
        },
    });
