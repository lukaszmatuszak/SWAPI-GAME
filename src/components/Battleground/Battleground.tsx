import React, { useEffect, useState } from 'react';
import { Button, Grid, Typography } from '@material-ui/core';
import { useTranslation } from 'react-i18next';

import DetailsTile from '../../components/DetailsTile/DetailsTile';
import { Starship } from '../../interfaces/Starship';
import { Character } from '../../interfaces/Character';
import { GameStatus } from '../../interfaces/GameStatus';
import { Status } from '../../interfaces/Status';
import { Challenger } from '../../interfaces/Challenger';
import { getRandomInteger } from '../../core/utils';
import { defineStatusForStarship, defineStatusForCharacter } from './utils';

const Battleground: React.FC<BattlegroundProps> = ({
    contestants,
}: BattlegroundProps) => {
    const { t } = useTranslation();

    const [gameStatus, setGameStatus] = useState<GameStatus>({
        playerOne: {
            name: null,
            status: null,
            mass: null,
            crew: null,
            score: 0,
        },
        playerTwo: {
            name: null,
            status: null,
            mass: null,
            crew: null,
            score: 0,
        },
    });

    const drawChallengers = (): Challenger[] => {
        const playerOne = getRandomInteger(contestants.length);
        const playerTwo = getRandomInteger(contestants.length);

        return [contestants[playerOne], contestants[playerTwo]];
    };

    const defineStatus = (
        firstChallenger: Challenger,
        secondChallenger: Challenger,
    ): Status[] => {
        if ((firstChallenger as Starship).crew) {
            return defineStatusForStarship(
                firstChallenger as Starship,
                secondChallenger as Starship,
            );
        }

        if ((firstChallenger as Character).mass) {
            return defineStatusForCharacter(
                firstChallenger as Character,
                secondChallenger as Character,
            );
        }
    };

    const prepareBattleground = (
        firstChallenger: Challenger,
        secondChallenger: Challenger,
    ): void => {
        const [statusPlayerOne, statusPlayerTwo] = defineStatus(
            firstChallenger,
            secondChallenger,
        );

        setGameStatus({
            playerOne: {
                name: firstChallenger.name,
                status: statusPlayerOne,
                mass: Number((firstChallenger as Character).mass),
                crew: Number((firstChallenger as Starship).crew),
                score:
                    statusPlayerOne !== Status.LOOSER
                        ? gameStatus.playerOne.score + 1
                        : gameStatus.playerOne.score,
            },
            playerTwo: {
                name: secondChallenger.name,
                status: statusPlayerTwo,
                mass: Number((secondChallenger as Character).mass),
                crew: Number((secondChallenger as Starship).crew),
                score:
                    statusPlayerTwo !== Status.LOOSER
                        ? gameStatus.playerTwo.score + 1
                        : gameStatus.playerTwo.score,
            },
        });
    };

    const startBattle = (): void => {
        const [firstChallenger, secondChallenger] = drawChallengers();

        prepareBattleground(firstChallenger, secondChallenger);
    };

    useEffect(() => {
        startBattle();
    }, [contestants]);

    return (
        <Grid xs={12} item container spacing={2} justify="center">
            <Grid container item xs={12} justify="center">
                <Typography variant="h3" data-test="score">
                    {gameStatus.playerOne.score} : {gameStatus.playerTwo.score}
                </Typography>
            </Grid>
            <Grid item xs={4}>
                <DetailsTile playerStatus={gameStatus.playerOne} />
            </Grid>
            <Grid item xs={4}>
                <DetailsTile playerStatus={gameStatus.playerTwo} />
            </Grid>
            <Grid container item xs={12} justify="center">
                <Button
                    variant="contained"
                    color="primary"
                    onClick={startBattle}
                    data-test="play-again"
                >
                    {t('PLAY_AGAIN')}
                </Button>
            </Grid>
        </Grid>
    );
};

interface BattlegroundProps {
    contestants: Challenger[];
}

export default Battleground;
