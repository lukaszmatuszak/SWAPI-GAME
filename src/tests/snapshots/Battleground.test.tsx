import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Battleground from '../../components/Battleground/Battleground';
import { contestantsStarshipMock } from '../fixtures/contestants';
import '../../core/i18n/i18n';

describe('Battleground Snapshot - ', () => {
    it('renders correctly', () => {
        const wrapper = shallow(
            <Battleground contestants={contestantsStarshipMock} />,
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
