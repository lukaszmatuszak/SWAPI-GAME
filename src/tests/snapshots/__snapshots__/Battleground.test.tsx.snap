// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Battleground Snapshot -  renders correctly 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  item={true}
  justify="center"
  spacing={2}
  xs={12}
>
  <WithStyles(ForwardRef(Grid))
    container={true}
    item={true}
    justify="center"
    xs={12}
  >
    <WithStyles(ForwardRef(Typography))
      data-test="score"
      variant="h3"
    >
      0
       : 
      0
    </WithStyles(ForwardRef(Typography))>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={4}
  >
    <DetailsTile
      playerStatus={
        Object {
          "crew": null,
          "mass": null,
          "name": null,
          "score": 0,
          "status": null,
        }
      }
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={4}
  >
    <DetailsTile
      playerStatus={
        Object {
          "crew": null,
          "mass": null,
          "name": null,
          "score": 0,
          "status": null,
        }
      }
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    item={true}
    justify="center"
    xs={12}
  >
    <WithStyles(ForwardRef(Button))
      color="primary"
      data-test="play-again"
      onClick={[Function]}
      variant="contained"
    >
      Play Again
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
